src/Strategy/Googlesource/RepoManifest.hs

```diff
@@ -0,0 +1,218 @@
+{-# language TemplateHaskell #-}
+
+module Strategy.Googlesource.RepoManifest
+  ( discover
+  , buildGraph
+  , analyze
+  , validatedProject
+  , validatedProjects
+  , nestedValidatedProjects
+
+  , RepoManifest(..)
+  , ManifestRemote(..)
+  , ManifestDefault(..)
+  , ManifestProject(..)
+  , ValidatedProject(..)
+
+  , mkProjectClosure
+  ) where
+
+import Prologue
+
+import Control.Carrier.Error.Either
+import qualified Data.Map.Strict as M
+import qualified Data.Text as T
+
+import DepTypes
+import Discovery.Walk
+import Effect.ReadFS
+import Graphing (Graphing, unfold)
+import Parse.XML
+import Types
+
+-- We're looking for a file called "manifest.xml" in a directory called ".repo"
+discover :: HasDiscover sig m => Path Abs Dir -> m ()
+discover = walk $ \_ subdirs files -> do
+  case find (\f -> (==) "manifest.xml" (fileName f)) files of
+    Nothing -> walkContinue
+    Just file -> do
+      if (dirname $ parent file) == $(mkRelDir ".repo") then do
+        runSimpleStrategy "googlesource-repomanifest" GooglesourceGroup $ analyze file
+        walkSkipAll subdirs
+      else walkContinue
+
+analyze :: (Has ReadFS sig m, Has (Error ReadFSErr) sig m, MonadFail m) => Path Rel File -> m ProjectClosureBody
+analyze file = do
+  projects <- nestedValidatedProjects file
+  pure $ mkProjectClosure file projects
+
+nestedValidatedProjects :: (Has ReadFS sig m, Has (Error ReadFSErr) sig m, MonadFail m) => Path Rel File -> m [ValidatedProject]
+nestedValidatedProjects file = do
+  manifest <- readContentsXML @RepoManifest file
+  validatedIncludedProjects <- validatedProjectsFromIncludes manifest $ parent file
+  let validatedDirectProjects = validatedProjects manifest
+  case validatedDirectProjects of
+    Nothing -> fail "Error"
+    Just ps -> pure $ ps ++ validatedIncludedProjects
+
+-- If a manifest has an include tag, the included manifest will be found in "manifests/<name attribute>" relative 
+-- to the original manifest file.
+-- A minimal manifest with an include tag will look like this:
+-- <manifest>
+--   <include name="default.xml" />
+-- </manifest>
+-- If you see that, you need to look for `manifests/default.xml`, where the manifests directory will
+-- be a sibling to the original manifest you were parsing.
+validatedProjectsFromIncludes :: (Has ReadFS sig m, Has (Error ReadFSErr) sig m, MonadFail m) => RepoManifest -> Path Rel Dir -> m [ValidatedProject]
+validatedProjectsFromIncludes manifest parentDir = do
+    let dirPath = toFilePath parentDir
+        manifestIncludeFiles :: [Text]
+        manifestIncludeFiles = map includeName $ manifestIncludes manifest
+        manifestFiles :: Maybe [Path Rel File]
+        manifestFiles = traverse (\file -> parseRelFile (dirPath ++ "manifests/" ++ T.unpack file)) manifestIncludeFiles
```

Comment:
can `manifestIncludeFiles` contain paths with `..`? `parseRelFile` doesn't like those.

you may actually want to use `resolveFile` from the `ReadFS` effect here anyway:
```hs
resolveFile :: (Has ReadFS sig m, Has (Error ReadFSErr) sig m) => Path Abs Dir -> Text -> m (Path Abs File)
```

You can do a similar `traverse`, and bind (`<-`) out the result

---

an alternative (though I generally prefer the option above) is to keep `parentDir` as a `Path Rel Dir`, and do something like this:
```hs
manifestFiles = traverse (\file -> (parentDir </>) <$> parseRelFile (T.unpack file)) manifestIncludeFiles
```

but maybe written a little more verbosely for readability

Reply:
`manifestIncludeFiles` cannot contain paths with `..`. As far as I can tell from the code (it's not really specified anywhere), it can only contain a filename without a directory -- the file it's pointing at must be in the `manifests` directory, and the `manifests` directory must be a sibling to the current `manifest.xml` that we're parsing.

If that's true, does that change your recommendation? I think I've only been dealing with `Rel` files, not `Abs` files, so I'm not sure if I can just drop in `resolveFile`.

Reply:
Yeah that definitely does change my recommendation -- something like the second option would be preferred in that case

Reply:
:+1:

Reply:
Okay, done. This looks much better too :)